extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/ShardTracker.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2020 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.jet.kinesis.impl;
+
+import com.amazonaws.services.kinesis.model.Shard;
+import com.hazelcast.jet.JetException;
+
+import java.math.BigInteger;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+
+public class ShardTracker {
+
+    /**
+     * We consider shards to be expired if they haven't been detected at all
+     * (neither as OPEN, nor as CLOSED) since at least this amount of time.
+     */
+    static final long EXPIRATION_MS = MINUTES.toMillis(10);
+
+    private final Map<String, TrackingInfo> info = new HashMap<>();
+    private final HashRange[] rangePartitions;
+
+    public ShardTracker(HashRange[] rangePartitions) {
+        this.rangePartitions = rangePartitions;
+    }
+
+    public void addUndetected(String shardId, BigInteger startingHashKey, long currentTimeMs) {
+        assert !info.containsKey(shardId);
+        info.put(shardId, new TrackingInfo(findOwner(startingHashKey), currentTimeMs));
+    }
+
+    public Map<Shard, Integer> markDetections(Set<Shard> shards, long currentTimeMs) {
+        Map<Shard, Integer> newShards = Collections.emptyMap();
+        for (Shard shard : shards) {
+            String shardId = shard.getShardId();
+            TrackingInfo trackingInfo = info.get(shardId);
+            if (trackingInfo == null) {
+                int owner = findOwner(new BigInteger(shard.getHashKeyRange().getStartingHashKey()));
+                trackingInfo = new TrackingInfo(owner, currentTimeMs);
+                info.put(shardId, trackingInfo);
+            }
+            boolean firstDetection = trackingInfo.markDetection(currentTimeMs);
+            if (firstDetection) {
+                if (newShards.isEmpty()) {
+                    newShards = new HashMap<>();
+                }
+                newShards.put(shard, trackingInfo.getOwner());
+            }
```

Comment:
Isn't it always the case that if `firstDetection == true`, then `trackingInfo` was null above? If yes, you can apply this suggestion:
```suggestion
            if (trackingInfo == null) {
                int owner = findOwner(new BigInteger(shard.getHashKeyRange().getStartingHashKey()));
                trackingInfo = new TrackingInfo(owner, currentTimeMs);
                info.put(shardId, trackingInfo);
                if (newShards.isEmpty()) {
                    newShards = new HashMap<>();
                }
                newShards.put(shard, trackingInfo.getOwner());
            }
            trackingInfo.markDetection(currentTimeMs);
```
From `TrackingInfo` you can then remove the `boolean detected` field.